File: Components/Strata.Foundation.Core/main/typescript/Utility/Expendable.ts
import {IOptional} from "./IOptional";
import {ExpendableContext} from "./ExpendableContext";
import {Optional} from "./Optional";
import {IConsumerOrLambda, LambdaConsumer} from "./LambdaConsumer";
import {IFunctionOrLambda, LambdaFunction} from "./LambdaFunction";
import {NoSuchElementException} from "./NoSuchElementException";
import {ISupplierOrLambda, LambdaSupplier} from "./LambdaSupplier";
import {IRunnableOrLambda, LambdaRunnable} from "./LambdaRunnable";
import {IPredicateOrLambda, LambdaPredicate} from "./LambdaPredicate";

export
class Expendable<T>
    implements IOptional<T>
{
    private readonly allowed: number;
    private          context: Optional<ExpendableContext<T>>;

    public constructor(value?:T, limit?: number)
    {
        if (value != null)
        {
            this.allowed = (limit != null && limit > 0) ? limit : 1;
            this.context = Optional.of(new ExpendableContext<T>(value, this.allowed));
        }
        else
        {
            this.allowed = 0;
            this.context = Optional.empty();
        }
    }

    public get(): T
    {
        return this.context
                .map<T>(context => this.apply(context,(value:T) => value))
                .orElseThrow(new NoSuchElementException("No value present"));
    }

    public ifPresent(consumer: IConsumerOrLambda<T>): void
    {
        this.context.ifPresent(context => this.accept(context, consumer));
    }

    public ifPresentOrElse<U>(present: IFunctionOrLambda<T,U>, notPresent: ISupplierOrLambda<U>): U
    {
        return this.context
            .map(context => this.apply(context, present))
            .orElseGet(LambdaSupplier.of(notPresent));
    }

    public ifPresentOrThrow<U>(present: IFunctionOrLambda<T,U>, exception: Error): U
    {
        return this.context
            .map(context => this.apply(context, present))
            .orElseThrow(exception);
    }

    public ifPresentOrElseNoReturn(
        present:    IConsumerOrLambda<T>,
        notPresent: IRunnableOrLambda): void
    {
        if (this.context.isPresent())
            this.context.ifPresent(context => this.accept(context, present));
        else
            LambdaRunnable.of(notPresent).run();
    }

    public ifPresentOrThrowNoReturn(
        present:    IConsumerOrLambda<T>,
        notPresent: Error): void
    {
        if (this.context.isPresent())
            this.context.ifPresent(context => this.accept(context, present));
        else
            throw notPresent;
    }

    public ifNotPresent(notPresent: IRunnableOrLambda): void
    {
        if (!this.context.isPresent())
            LambdaRunnable.of(notPresent).run();
    }

    public filter(predicate: IPredicateOrLambda<T>): Expendable<T>
    {
        return this.context
            .filter(context => LambdaPredicate.of(predicate).test(context.getValue()))
            .map(context => new Expendable<T>(context.getValue(), this.allowed))
            .orElse(Expendable.empty());
    }

    public map<U>(mapper: IFunctionOrLambda<T,U>): Expendable<U>
    {
        return this.context
            .map(context => new Expendable<U>(this.apply(context, mapper), this.allowed))
            .orElse(Expendable.empty());
    }

    public flatMap<U>(mapper: IFunctionOrLambda<T,IOptional<U>>): Expendable<U>
    {
        return this.context
            .map(context => this.apply(context, mapper))
            .orElse(Expendable.empty()) as Expendable<U>;
    }

    or(supplier: ISupplierOrLambda<IOptional<T>>): Expendable<T>
    {
        if (this.isPresent())
            return this;

        return LambdaSupplier.of(supplier).get() as Expendable<T>;
    }

    public orElse(other: T): T
    {
        return this.context
            .map(context => this.apply(context,(value:T) => value))
            .orElse(other);
    }

    public orElseGet(other: ISupplierOrLambda<T>): T
    {
        return this.context
            .map(context => this.apply(context,(value:T) => value))
            .orElseGet(LambdaSupplier.of(other));
    }

    public orElseThrow(exception: Error): T
    {
        return this.context
            .map(context => this.apply(context,(value:T) => value))
            .orElseThrow(exception);
    }

    orElseGetThrow(supplier: ISupplierOrLambda<Error>): T
    {
        return this.context
            .map(context => this.apply(context,(value:T) => value))
            .orElseGetThrow(supplier);
    }

    public isPresent(): boolean
    {
        return this.context.isPresent();
    }

    public isEmpty(): boolean
    {
        return this.context.isEmpty();
    }

    public isExpended(): boolean
    {
        return this.context
            .map(context => context.isExpended())
            .orElse(true);
    }

    public getAllowed(): number
    {
        return this.allowed;
    }

    public getRemaining(): number
    {
        return this.context
            .map(context => context.getRemaining())
            .orElse(0);
    }

    public static of<T>(value?: T, limit?: number): Expendable<T>
    {
        return new Expendable<T>(value, limit);
    }

    public static empty<T>(limit?: number): Expendable<T>
    {
        return new Expendable<T>(undefined, limit);
    }


    private accept(context: ExpendableContext<T>, consumer: IConsumerOrLambda<T>): void
    {
        LambdaConsumer.of(consumer).accept(context.getValue());
        context.decrementRemaining();

        if (context.isExpended())
            this.context = Optional.empty();
    }

    private apply<U>(context: ExpendableContext<T>, func: IFunctionOrLambda<T,U>): U
    {
        const result: U = LambdaFunction.of(func).apply(context.getValue());
        context.decrementRemaining();

        if (context.isExpended())
            this.context = Optional.empty();

        return result;
    }
}

File: Components/Strata.Foundation.Core/main/java/strata/foundation/core/utility/Expendable.java
//////////////////////////////////////////////////////////////////////////////
// Expendable.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.utility;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

/**
 * <p>
 * {@link IOptionable} implementation that wraps a value which may be
 * retrieved only a limited number of times before it is automatically
 * treated as empty, useful for representing single-use or rate-limited
 * resources.
 * </p>
 * <br/>
 * <b>Type Parameter</b><br/>
 * {@code <T>} - the type of the wrapped value
 * <br/>
 * <b>Examples</b><br/>
 * <pre>
 * Expendable&lt;String&gt; token = Expendable.of("one-time-token",1);
 *
 * token.ifPresent(t -&gt; authenticate(t));
 * </pre>
 */
public
class Expendable<T>
    implements IOptionable<T>
{
    private final int                      allowed;
    private Optional<ExpendableContext<T>> context;

    public
    Expendable()
    {
        this(null,0);
    }

    public
    Expendable(T value)
    {
        this(value,1);
    }

    public
    Expendable(T value,int allowed)
    {
        if (value != null)
        {
            this.allowed = Math.max(1,allowed);
            this.context = Optional.of(new ExpendableContext<>(value,this.allowed));
        }
        else
        {
            this.allowed = 0;
            this.context = Optional.empty();
        }
    }

    @Override
    public int
    hashCode()
    {
        return Objects.hash(allowed,context);
    }

    @Override
    public boolean
    equals(Object other)
    {
        return
            (other instanceof Expendable) &&
            Objects.equals(allowed,((Expendable<?>)other).allowed) &&
            Objects.equals(context,((Expendable<?>)other).context);
    }

    @Override
    public T
    get()
    {
        return
            context
                .map(c -> this.apply(c,value -> value))
                .orElseThrow();
    }

    @Override
    public void
    ifPresent(Consumer<T> consumer)
    {
        context.ifPresent(c -> accept(c,consumer));
    }

    @Override
    public <U> U
    ifPresentOrElse(Function<T,U> present,Supplier<U> notPresent)
    {
        return
            context
                .map(c -> this.apply(c,present))
                .orElseGet(notPresent);
    }

    @Override
    public <U,E extends RuntimeException> U
    ifPresentOrThrow(Function<T,U> present,E exception)
        throws E
    {
        return
            context
                .map(c -> this.apply(c,present))
                .orElseThrow(() -> exception);
    }

    @Override
    public void
    ifPresentOrElseNoReturn(Consumer<T> present,Runnable notPresent)
    {
        if (context.isPresent())
            context.ifPresent(c -> this.accept(c,present));
        else
            notPresent.run();
    }

    @Override
    public <E extends RuntimeException> void
    ifPresentOrThrowNoReturn(Consumer<T> present,E exception)
        throws E
    {
        if (context.isPresent())
            context.ifPresent(c -> this.accept(c,present));
        else
            throw exception;
    }

    @Override
    public void
    ifEmpty(Runnable notPresent)
    {
        if (context.isEmpty())
            notPresent.run();
    }

    @Override
    public Expendable<T>
    filter(Predicate<T> predicate)
    {
        return
            context
                .filter(c -> predicate.test(c.getValue()))
                .map(c -> new Expendable<>(c.getValue(),allowed))
                .orElseGet(() -> Expendable.empty());
    }

    @Override
    public <U> Expendable<U>
    map(Function<T,U> mapper)
    {
        return
            context
                .map(c -> new Expendable<>(this.apply(c,mapper),allowed))
                .orElseGet(() -> Expendable.empty());
    }

    @Override
    public <U,O extends IOptionable<U>> Expendable<U>
    flatMap(Function<T,O> mapper)
    {
        return
            context
                .map(c -> this.applyForFlatMap(c,mapper))
                .orElse(Expendable.empty());
    }

    @Override
    public <O extends IOptionable<T>> Expendable<T>
    or(Supplier<O> supplier)
    {
        Objects.requireNonNull(supplier, "Supplier must not be null");

        return
            context.isPresent()
                ? this
                : new Expendable<>(
                    supplier
                        .get()
                        .orElse(null),
                    allowed);
    }

    @Override
    public T
    orElse(T other)
    {
        return
            context
                .map(c -> this.apply(c,value -> value))
                .orElse(other);
    }

    @Override
    public T
    orElseGet(Supplier<T> other)
    {
        return
            context
                .map(c -> this.apply(c,value -> value))
                .orElseGet(other);
    }

    @Override
    public T
    orElseThrow() throws NoSuchElementException
    {
        return
            context
                .map(c -> this.apply(c,value -> value))
                .orElseThrow();
    }

    @Override
    public <E extends RuntimeException> T
    orElseThrow(Supplier<E> exceptionSupplier)
    {
        return
            context
                .map(c -> this.apply(c,value -> value))
                .orElseThrow(exceptionSupplier);
    }

    @Override
    public boolean
    isPresent()
    {
        return context.isPresent();
    }

    @Override
    public boolean
    isEmpty()
    {
        return context.isEmpty();
    }

    @Override
    public Optional<T> toOptional()
    {
        return Optional.empty();
    }

    public boolean
    isExpended()
    {
        return
            context
                .orElse(new ExpendableContext<>(null,0))
                .isExpended();
    }

    public int
    getAllowed() { return allowed; }

    public int
    getRemaining()
    {
        return
            context
                .map(ExpendableContext::getRemaining)
                .orElse(0);
    }

    private void
    writeObject(ObjectOutputStream out)
        throws IOException
    {
        if (context.isPresent())
        {
            out.writeBoolean(true);
            out.writeObject(context.get());
        }
        else
            out.writeBoolean(false);
    }

    private void
    readObject(ObjectInputStream in)
        throws IOException,ClassNotFoundException
    {
        boolean isPresent = in.readBoolean();

        if (isPresent)
            context = Optional.ofNullable((ExpendableContext<T>)in.readObject());
        else
            context = Optional.empty();
    }

    public static <T> Expendable<T>
    of(T value)
    {
        return of(value,1);
    }

    public static <T> Expendable<T>
    of(T value,int allowed)
    {
        return new Expendable<>(value,allowed);
    }

    public static <T> Expendable<T>
    empty()
    {
        return of(null,0);
    }

    private void
    accept(ExpendableContext<T> context,Consumer<T> consumer)
    {
        consumer.accept(context.getValue());
        context.decrementRemaining();

        if (context.isExpended())
            this.context = Optional.empty();
    }


    private <U> U
    apply(ExpendableContext<T> context,Function<T,U> function)
    {
        U output = function.apply(context.getValue());

        context.decrementRemaining();

        if (context.isExpended())
            this.context = Optional.empty();

        return output;
    }

    private <U> Expendable<U>
    applyForFlatMap(
        ExpendableContext<T>               context,
        Function<T,? extends IOptionable<U>> function)
    {
        Expendable<U> output = (Expendable<U>)function.apply(context.getValue());

        context.decrementRemaining();

        if (context.isExpended())
            this.context = Optional.empty();

        return output;
    }

}

//////////////////////////////////////////////////////////////////////////////
